[//]: # (This file is automatically generated. If you wish to make any changes, please use the JSON files and regenerate this file using the tpframework.)

# Try Catch Finally

Tags: sast, php, php_v7.4.9

Version: v1.0

## Description

In PHP [Exceptions](https://www.php.net/manual/en/language.exceptions.php) are similar to other languages. An exception can be thrown and caught, try blocks indicate positions in code, where exceptions can occur. This pattern should target the try-catch-finally block. The `finally` block will always be executed, regardless of whether an exception has been thrown, and before normal execution resumes.

## Overview

| Instances                 | has discovery rule   | discovery method   | rule successfull   |
|---------------------------|----------------------|--------------------|--------------------|
| [1 Instance](#1-instance) | yes                  | joern              | yes                |
| [2 Instance](#2-instance) | yes                  | joern              | yes                |

<details markdown="1"open>
<summary>

## 1 Instance
</summary>

This instance shows a 'try, catch, finally' construct. The unsanitized `$bar` from the `catch` will be returned.

### Code

```PHP
<?php
function foo($b) {
    $bar = $b;
    try {
        throw new Exception('Exception');
    } catch (Exception $e) {
        echo "catch \n";
        // Before return, the finally will run
        return $bar;
    } finally {
        echo "finally \n";
        //sanitize_xss
        $bar = htmlspecialchars($bar);
    }
}
$b = $_GET["p1"]; // source
$a = foo($b . "\n");
echo $a; // sink
// The output:
// catch
// finally
// $b without sanitize, XSS vulnerability
```

### Instance Properties

| category   | feature_vs_internal_api   | input_sanitizer   | negative_test_case   | source_and_sink   |
|------------|---------------------------|-------------------|----------------------|-------------------|
| D2         | FEATURE                   | no                | no                   | no                |

<details markdown="1">
<summary>
<b>More</b></summary>

<details markdown="1">
<summary>

### Compile
</summary>

```bash
$_main:
     ; (lines=10, args=0, vars=2, tmps=6)
     ; (before optimizer)
     ; /.../PHP/52_try_catch_finally/1_instance_52_try_catch_finally/1_instance_52_try_catch_finally.php:1-22
     ; return  [] RANGE[0..0]
0000 T2 = FETCH_R (global) string("_GET")
0001 T3 = FETCH_DIM_R T2 string("p1")
0002 ASSIGN CV0($b) T3
0003 INIT_FCALL 1 240 string("foo")
0004 T5 = CONCAT CV0($b) string("
")
0005 SEND_VAL T5 1
0006 V6 = DO_UCALL
0007 ASSIGN CV1($a) V6
0008 ECHO CV1($a)
0009 RETURN int(1)

foo:
     ; (lines=21, args=1, vars=3, tmps=7)
     ; (before optimizer)
     ; /.../PHP/52_try_catch_finally/1_instance_52_try_catch_finally/1_instance_52_try_catch_finally.php:2-15
     ; return  [] RANGE[0..0]
0000 CV0($b) = RECV 1
0001 ASSIGN CV1($bar) CV0($b)
0002 V5 = NEW 1 string("Exception")
0003 SEND_VAL_EX string("Exception") 1
0004 DO_FCALL
0005 THROW V5
0006 JMP 0012
0007 CV2($e) = CATCH string("Exception")
0008 ECHO string("catch 
")
0009 T7 = QM_ASSIGN CV1($bar)
0010 T4 = FAST_CALL 0014 T7
0011 RETURN T7
0012 T4 = FAST_CALL 0014
0013 JMP 0020
0014 ECHO string("finally 
")
0015 INIT_FCALL 1 96 string("htmlspecialchars")
0016 SEND_VAR CV1($bar) 1
0017 V8 = DO_ICALL
0018 ASSIGN CV1($bar) V8
0019 FAST_RET T4
0020 RETURN null
LIVE RANGES:
     5: 0003 - 0005 (new)
     7: 0010 - 0011 (tmp/var)
EXCEPTION TABLE:
     0002, 0007, 0014, 0019
```

</details>

<details markdown="1">
<summary>

### Discovery
</summary>

The rule searches for `FAST_RET` on opcode level.

```scala
val x52 = (name, "52_try_catch_finally_iall", cpg.call(".*FAST_RET.*").location.toJson);
```

| discovery method   | expected accuracy   |
|--------------------|---------------------|
| joern              | Perfect             |

</details>

<details markdown="1"open>
<summary>

### Measurement
</summary>

| Tool        | Comm_1   | Comm_2   | phpSAFE   | Progpilot   | RIPS   | WAP   | Ground Truth   |
|-------------|----------|----------|-----------|-------------|--------|-------|----------------|
| 08 Jun 2021 | yes      | yes      | no        | no          | yes    | no    | yes            |
| 22 May 2023 | yes      | yes      |           |             |        |       | yes            |

</details>

</details>

</details>

<details markdown="1">
<summary>

## 2 Instance
</summary>

This instance shows the same construct as the previous instance. But this time, the sanitized value from the finally statement is returned.

### Code

```PHP
<?php
function foo($b) {
    $bar = $b;
    try {
        throw new Exception('Exception');
    } catch (Exception $e) {
        echo "catch \n";
        // Before return, finally will run
        return $bar;
    } finally {
        echo "finally \n";
        //sanitize_xss
        $bar = htmlspecialchars($bar);
        // This value will return from the function
        return $bar;
    }
}
$b = $_GET["p1"]; // source
$a = foo($b . "\n");
echo $a; // sink
// The output:
// catch
// finally
// $b with sanitize, No vulnerability
```

### Instance Properties

| category   | feature_vs_internal_api   | input_sanitizer   | negative_test_case   | source_and_sink   |
|------------|---------------------------|-------------------|----------------------|-------------------|
| D2         | FEATURE                   | no                | yes                  | no                |

<details markdown="1">
<summary>
<b>More</b></summary>

<details markdown="1">
<summary>

### Compile
</summary>

```bash
$_main:
     ; (lines=10, args=0, vars=2, tmps=6)
     ; (before optimizer)
     ; /.../PHP/52_try_catch_finally/2_instance_52_try_catch_finally/2_instance_52_try_catch_finally.php:1-24
     ; return  [] RANGE[0..0]
0000 T2 = FETCH_R (global) string("_GET")
0001 T3 = FETCH_DIM_R T2 string("p1")
0002 ASSIGN CV0($b) T3
0003 INIT_FCALL 1 240 string("foo")
0004 T5 = CONCAT CV0($b) string("
")
0005 SEND_VAL T5 1
0006 V6 = DO_UCALL
0007 ASSIGN CV1($a) V6
0008 ECHO CV1($a)
0009 RETURN int(1)

foo:
     ; (lines=23, args=1, vars=3, tmps=7)
     ; (before optimizer)
     ; /.../PHP/52_try_catch_finally/2_instance_52_try_catch_finally/2_instance_52_try_catch_finally.php:2-17
     ; return  [] RANGE[0..0]
0000 CV0($b) = RECV 1
0001 ASSIGN CV1($bar) CV0($b)
0002 V5 = NEW 1 string("Exception")
0003 SEND_VAL_EX string("Exception") 1
0004 DO_FCALL
0005 THROW V5
0006 JMP 0012
0007 CV2($e) = CATCH string("Exception")
0008 ECHO string("catch 
")
0009 T7 = QM_ASSIGN CV1($bar)
0010 T4 = FAST_CALL 0014 T7
0011 RETURN T7
0012 T4 = FAST_CALL 0014
0013 JMP 0022
0014 ECHO string("finally 
")
0015 INIT_FCALL 1 96 string("htmlspecialchars")
0016 SEND_VAR CV1($bar) 1
0017 V8 = DO_ICALL
0018 ASSIGN CV1($bar) V8
0019 DISCARD_EXCEPTION T4
0020 RETURN CV1($bar)
0021 FAST_RET T4
0022 RETURN null
LIVE RANGES:
     5: 0003 - 0005 (new)
     7: 0010 - 0011 (tmp/var)
EXCEPTION TABLE:
     0002, 0007, 0014, 0021
```

</details>

<details markdown="1">
<summary>

### Discovery
</summary>

The rule searches for `FAST_RET` on opcode level.

```scala
val x52 = (name, "52_try_catch_finally_iall", cpg.call(".*FAST_RET.*").location.toJson);
```

| discovery method   | expected accuracy   |
|--------------------|---------------------|
| joern              | Perfect             |

</details>

<details markdown="1"open>
<summary>

### Measurement
</summary>

| Tool        | Comm_1   | Comm_2   | RIPS   | Ground Truth   |
|-------------|----------|----------|--------|----------------|
| 08 Jun 2021 | yes      | yes      | no     | no             |
| 22 May 2023 | yes      | yes      |        | no             |

</details>

</details>

</details>
